### PR TITLE
Virtual Destructors and Callback Function

### PR DESCRIPTION
**Pull request by PatWie**
This is essentially a fix for the issue #37. @jdumas, @spinicist , can you have a look at the code? There is now a virtual callback function which can decide within the problem class whether the optimization should continue or not.

In addition simple_withoptions.cpp has an example how to fetch and print the current state.
In addition addition, I removed the config file for cmake, since the usage of this was probably a bad idea.
In addition addition addition, the streaming function in the meta file is now declared as inline.

### CODE REVIEW

**Comment by jdumas** [comment]
Hum, why take a `Vector<T> x` as an argument and not a `const Vector<T> &x`<sup>[1](#hacknote)</sup>? I know recent compilers optimize pass-by-value, but I don't see how it would avoid the explicit copy here. (The same goes with the methods `lowerBound()` / `upperBound` btw, it should return a `const &`).

Why does the `callback()` method takes a `Criteria<> *` and not a `Criteria<> &`? Can the argument be a `nullptr`?

Last remark, I would formulate the stop condition as:

    while (objFunc.callback(&this->m_current, x0) && (this->m_status == Status::Continue))

Instead of:

    while ((this->m_status == Status::Continue) && objFunc.callback(&this->m_current, x0))

Which allows the last line to print the final stats corresponding to the value of `x0` returned by the optimizer.

-----
<sub><a name="hacknote">1</a>: You could even pass it as a `Vector<T> &x`, which could allow for certain hacks like renormalizing your rotations between each iterations, etc. (But whether or not this should be a supported behavior is more debatable).</sub>

File: .travis.yml [hunk 1]
```diff
@@ -62,7 +62,7 @@ install:
 script:
   - mkdir build
   - cd build
-  - cmake -DCMAKE_CXX_COMPILER=$COMPILER ..
+  - EIGEN3_INCLUDE_DIR=eigen cmake -DCMAKE_CXX_COMPILER=$COMPILER ..
```
Comment (spinicist) [inline .travis.yml]: I think this line should read:
`cmake -DCMAKE_CXX_COMPILER=$COMPILER -DEIGEN3_INCLUDE_DIR=eigen`

**Comment by spinicist** [comment]
Hi,
I agree with @jdumas comments. `callback()` should take a `const Criteria<> &` instead of a pointer.

One further issue is how should `callback()` interact with the Debug print infrastructure I started on in my last PR? I didn't get very far with it, and I've come round to thinking that `callback()` is more elegant. So either the Debug stuff should get removed or should be implemented on top of `callback()`.

Finally - I made a note on @8a999bd - I think the Eigen include directory has to be passed with a `-D` option.

**Comment by PatWie** [comment]
And now?

**Comment by jdumas** [comment]
Seems legit to me =)

**Pull request merged**